### ISSUE TITLE
Abstract the dagu's storage layer

### ISSUE THREAD

**Issue by Arvintian** [yohamta/dagu#267]
Base on executor interface, dagu can scheduler any task. But dagu's storage rely on local file system. It is limit dagu deploy HA mode, and scheduler massive DAG task. I think we can abstract the dagu's storage layer as interface. Like [chartmuseum's storage](https://github.com/chartmuseum/storage), may be we can direct use chartmuseum's library. Let dagu run on cloud !

**Comment by yohamta** [comment]
Thanks for the great suggestion as always :) 

> It is limit dagu deploy as HA mode, and scheduler massive DAG task

Yeah, that's absolutely true. I quickly searched and found some other modules that might be useful.
- https://github.com/chartmuseum/storage
- https://github.com/Shopify/go-storage
- https://github.com/beyondstorage/go-storage

I was vaguely thinking of using shared mounted NFS (EFS, etc.) as the storage for reading and writing status files and logs, but as you say, it would be great if users also could use OSS (S3, GCS, etc) as the storage (with some caching mechanism to avoid unnecessary network cost).

**Comment by Arvintian** [comment]
Yes, use shared mounted NFS is ok to save dagu's data. I am try use dagu to build a devops paltform, mange object storage is easier than nfs. But it is ok, caching mechanism is important about performance. This feature we can do after and think more.

**Comment by yohamta** [comment]
Yeah, will do abstraction first, we will think how we can utilize object storage :)

**Comment by yohamta** [comment]
This might be useful for storage abstraction.
[https://github.com/C2FO/vfs](https://github.com/C2FO/vfs)

**Comment by yohamta0** [comment]
Closing the storage-abstraction portion as completed by #2194. Persistence now uses pluggable Backend and Collection interfaces instead of tying stores directly to a single implementation.

Support for a specific remote object-storage backend such as S3 or GCS is a separate feature and should be tracked in a focused issue if it is still desired.
